pkg/psmdb/mongos.go

```diff
@@ -0,0 +1,361 @@
+	}
+
+	return inits
+}
+
+func mongosContainer(cr *api.PerconaServerMongoDB) (corev1.Container, error) {
+	fvar := false
+
+	resources, err := CreateResources(cr.Spec.Sharding.Mongos.ResourcesSpec)
+	if err != nil {
+		return corev1.Container{}, fmt.Errorf("resource creation: %v", err)
+	}
+
+	volumes := []corev1.VolumeMount{
+		{
+			Name:      MongodDataVolClaimName,
+			MountPath: MongodContainerDataDir,
+		},
+		{
+			Name:      InternalKey(cr),
+			MountPath: mongodSecretsDir,
+			ReadOnly:  true,
+		},
+		{
+			Name:      "ssl",
+			MountPath: sslDir,
+			ReadOnly:  true,
+		},
+		{
+			Name:      "ssl-internal",
+			MountPath: sslInternalDir,
+			ReadOnly:  true,
+		},
+	}
+
+	if *cr.Spec.Mongod.Security.EnableEncryption {
+		volumes = append(volumes,
+			corev1.VolumeMount{
+				Name:      cr.Spec.Mongod.Security.EncryptionKeySecret,
+				MountPath: mongodRESTencryptDir,
+				ReadOnly:  true,
+			},
+		)
+	}
+
+	container := corev1.Container{
+		Name:            "mongos",
+		Image:           cr.Spec.Image,
+		ImagePullPolicy: cr.Spec.ImagePullPolicy,
+		Args:            mongosContainerArgs(cr, resources),
+		Ports: []corev1.ContainerPort{
+			{
+				Name:          mongosPortName,
+				HostPort:      cr.Spec.Sharding.Mongos.HostPort,
+				ContainerPort: cr.Spec.Sharding.Mongos.Port,
+			},
+		},
+		Env: []corev1.EnvVar{
+			{
+				Name:  "SERVICE_NAME",
+				Value: cr.Name,
+			},
+			{
+				Name:  "NAMESPACE",
+				Value: cr.Namespace,
+			},
+			{
+				Name:  "MONGODB_PORT",
+				Value: strconv.Itoa(int(cr.Spec.Sharding.Mongos.Port)),
+			},
+		},
+		EnvFrom: []corev1.EnvFromSource{
+			{
+				SecretRef: &corev1.SecretEnvSource{
+					LocalObjectReference: corev1.LocalObjectReference{
+						Name: cr.Spec.Secrets.Users,
+					},
+					Optional: &fvar,
+				},
+			},
+		},
+		WorkingDir:      MongodContainerDataDir,
+		LivenessProbe:   &cr.Spec.Sharding.Mongos.LivenessProbe.Probe,
+		ReadinessProbe:  cr.Spec.Sharding.Mongos.ReadinessProbe,
+		SecurityContext: cr.Spec.Sharding.Mongos.ContainerSecurityContext,
+		Resources:       resources,
+		VolumeMounts:    volumes,
+	}
+
+	if cr.CompareVersion("1.5.0") >= 0 {
+		container.EnvFrom = []corev1.EnvFromSource{
+			{
+				SecretRef: &corev1.SecretEnvSource{
+					LocalObjectReference: corev1.LocalObjectReference{
+						Name: "internal-" + cr.Name + "-users",
+					},
+					Optional: &fvar,
+				},
+			},
+		}
+		container.Command = []string{"/data/db/ps-entry.sh"}
+	}
+
+	return container, nil
+}
+
+func mongosContainerArgs(cr *api.PerconaServerMongoDB, resources corev1.ResourceRequirements) []string {
+	mdSpec := cr.Spec.Mongod
+	msSpec := cr.Spec.Sharding.Mongos
+	cfgRs := cr.Spec.Sharding.ConfigsvrReplSet
+
+	cfgInstanses := make([]string, 0, cfgRs.Size)
+	for i := 0; i < int(cfgRs.Size); i++ {
+		cfgInstanses = append(cfgInstanses,
+			fmt.Sprintf("%s-%s-%d.%s-%s.%s.svc.cluster.local:%d",
+				cr.Name, cfgRs.Name, i, cr.Name, cfgRs.Name, cr.Namespace, msSpec.Port))
```

Comment:
don't use list, please use `mongodb+srv://` connection string with service name

Reply:
Looks like mongos can't parse connection string for config server in such format
```
exec mongos ... --configdb mongodb+srv://my-cluster-name-cfg.run-mongos-1405.svc.cluster.local/?replicaSet=cfg
Failed global initialization: FailedToParse: Must specify set name for replica set ConnectionStrings
```
And based on documentation it should not
https://docs.mongodb.com/manual/reference/program/mongos/#cmdoption-mongos-configdb